roxie/ccd/ccdfile.cpp

```diff
@@ -1275,28 +1276,32 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
             closePending[remote] = false;
         }
         CriticalBlock b(crit);
-        ICopyArrayOf<ILazyFileIO> goers;
+        IArrayOf<ILazyFileIO> goers;
         HashIterator h(files);
         ForEach(h)
         {
-            ILazyFileIO *f = files.mapToValue(&h.query());
-            if (f->isAlive() && f->isOpen() && f->isRemote()==remote && !f->isCopying())
+            ILazyFileIO * match = files.mapToValue(&h.query());
+            if (match->isAliveAndLink())
             {
-                unsigned age = msTick() - f->getLastAccessed();
-                if (age > maxFileAge[remote])
+                Owned<ILazyFileIO> f = match;
+                if (f->isOpen() && f->isRemote()==remote && !f->isCopying())
                 {
-                    if (traceLevel > 5)
+                    unsigned age = msTick() - f->getLastAccessed();
+                    if (age > maxFileAge[remote])
                     {
-                        // NOTE - querySource will cause the file to be opened if not already open
-                        // That's OK here, since we know the file is open and remote.
-                        // But don't be tempted to move this line outside these if's (eg. to trace the idle case)
-                        const char *fname = remote ? f->querySource()->queryFilename() : f->queryFilename();
-                        DBGLOG("Closing inactive %s file %s (last accessed %u ms ago)", remote ? "remote" : "local",  fname, age);
+                        if (traceLevel > 5)
+                        {
+                            // NOTE - querySource will cause the file to be opened if not already open
+                            // That's OK here, since we know the file is open and remote.
+                            // But don't be tempted to move this line outside these if's (eg. to trace the idle case)
+                            const char *fname = remote ? f->querySource()->queryFilename() : f->queryFilename();
+                            DBGLOG("Closing inactive %s file %s (last accessed %u ms ago)", remote ? "remote" : "local",  fname, age);
+                        }
+                        f->close();
                     }
-                    f->close();
+                    else
+                        goers.append(*f.getClear());
                 }
-                else
-                    goers.append(*f);
             }
         }
         unsigned numFilesLeft = goers.ordinality(); 
```

Comment:
@richardkchapman if goers are now linked could the critical section finish here instead?

Reply:
I would have thought so